cli/src/commands/git.rs

```diff
@@ -1018,6 +1043,104 @@ fn cmd_git_submodule_print_gitmodules(
+    let submodule_path = tx
+        .mut_repo()
+        .submodule_store()
+        .get_submodule_path(&args.name);
+    let submodule_path_existed = submodule_path.exists();
+    if submodule_path_existed {
+        if !is_empty_dir(&submodule_path) {
+            return Err(user_error(format!(
+                "Could not clone submodule: {} exists",
+                submodule_path.display()
+            )));
+        }
+    } else {
+        fs::create_dir(&submodule_path).unwrap();
+    }
+
+    let clone_result = do_submodule_clone(ui, command, &submodule_path, &submodule.url);
+    if clone_result.is_err() {
+        // Canonicalize because fs::remove_dir_all() doesn't seem to like e.g.
+        // `/some/path/.`
+        let canonical_submodule_path = submodule_path.canonicalize().unwrap();
+        if let Err(err) = fs::remove_dir_all(canonical_submodule_path.join(".jj")).and_then(|_| {
+            if !submodule_path_existed {
+                fs::remove_dir(&canonical_submodule_path)
+            } else {
+                Ok(())
+            }
+        }) {
+            writeln!(
+                ui,
+                "Failed to clean up {}: {}",
+                canonical_submodule_path.display(),
+                err
+            )
+            .ok();
+        }
+    }
+    writeln!(ui, "Cloned submodule \"{}\"", submodule.name)?;
+    // Unlike cmd_git_clone(), don't update the submodule's working copy.
+    // TODO do we need this transaction? We aren't changing branches
+    tx.finish(ui)?;
+    Ok(())
+}
+
+// This is mostly copied from do_git_clone(). We might want to dedup them in the
+// future.
+fn do_submodule_clone(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    submodule_path: &Path,
+    submodule_url: &str,
+) -> Result<(), CommandError> {
+    // TODO The submodule shouldn't have a workspace, but we don't have a good
+    // way to init a repo without one.
+    let (workspace, repo) = Workspace::init_internal_git(command.settings(), submodule_path)?;
+    let git_repo = get_git_repo(repo.store())?;
+    let mut submodule_command = command.for_loaded_repo(ui, workspace, repo)?;
+    let remote_name = "origin";
+    git_repo.remote(remote_name, submodule_url).unwrap();
+    let mut fetch_tx = submodule_command.start_transaction("fetch from git remote into empty repo");
+
+    with_remote_callbacks(ui, |cb| {
+        git::fetch(
+            fetch_tx.mut_repo(),
+            &git_repo,
+            remote_name,
+            None,
+            cb,
+            &command.settings().git_settings(),
+        )
+    })
+    .map_err(|err| match err {
+        GitFetchError::NoSuchRemote(_) => {
+            panic!("shouldn't happen as we just created the git remote")
+        }
+        GitFetchError::InternalGitError(err) => map_git_error(err),
+        GitFetchError::InvalidGlob => {
+            unreachable!("we didn't provide any globs")
+        }
+    })?;
+    fetch_tx.finish(ui)?;
```

Comment:
When you have a regular `Transaction` object (not this CLI-level wrapper), you'll get a new `ReadonlyRepo` back when you commit it. Since no one uses that yet, you can just ignore it.